### PR TITLE
made time and timezone optional

### PR DESCRIPTION
**Pull request by edewit**
fixes: #134

### CODE REVIEW

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java [hunk 1]
```diff
@@ -163,8 +166,13 @@ private static Consumer<Path> addBuildItem(Path root, HashSet<RoqFrontMatterRawT
...
                     String dateString = null;
                     if (fm.containsKey(DATE_KEY)) {
-                        ZonedDateTime date = ZonedDateTime.parse(fm.getString(DATE_KEY),
-                                DateTimeFormatter.ofPattern(config.dateFormat()));
+                        ZonedDateTime date = new DateTimeFormatterBuilder().appendPattern("yyyy-MM-dd[ HH:mm:ss Z]")
+                                .parseDefaulting(ChronoField.HOUR_OF_DAY, 12)
+                                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+                                .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+                                .toFormatter()
+                                .withZone(ZoneId.of("UTC"))
```
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: We should use the local timezone by default no?

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java [hunk 2]
```diff
@@ -163,8 +166,13 @@ private static Consumer<Path> addBuildItem(Path root, HashSet<RoqFrontMatterRawT
                     final String content = stripFrontMatter(fullContent);
                     String dateString = null;
                     if (fm.containsKey(DATE_KEY)) {
-                        ZonedDateTime date = ZonedDateTime.parse(fm.getString(DATE_KEY),
-                                DateTimeFormatter.ofPattern(config.dateFormat()));
+                        ZonedDateTime date = new DateTimeFormatterBuilder().appendPattern("yyyy-MM-dd[ HH:mm:ss Z]")
```
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: Use `config.dateFormat()` and replace the default value there

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java [hunk 3]
```diff
@@ -163,8 +166,13 @@ private static Consumer<Path> addBuildItem(Path root, HashSet<RoqFrontMatterRawT
...
+                                .parseDefaulting(ChronoField.HOUR_OF_DAY, 12)
+                                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+                                .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+                                .toFormatter()
+                                .withZone(ZoneId.of("UTC"))
+                                .parse(fm.getString(DATE_KEY), ZonedDateTime::from);
```
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: Here we should use the filename date as default value (if present)

Comment (ia3andy) [reply roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: for example `2024-08-29-welcome-to-roq.md`

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java (hunk 1 above)
Comment (edewit) [reply roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: Yes, this only comes into play when there is no timzone set

Comment (ia3andy) [reply roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: Yes, but I mean when not set, we should use the system timezone (not UTC)

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java (hunk 3 above)
Comment (edewit) [reply roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: That wasn't here originally right? That is something new you want to introduce?

Comment (ia3andy) [reply roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/scan/RoqFrontMatterScanProcessor.java]: Yes

**Comment by melloware** [comment]
@all-contributors add @edewit for code

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java [hunk 4]
```diff
@@ -49,9 +49,15 @@ public interface RoqFrontMatterConfig {
...
      */
-    @WithDefault("yyyy-MM-dd HH:mm:ss Z")
+    @WithDefault("yyyy-MM-dd[ HH:mm][:ss][ Z]")
     String dateFormat();
 
+    /**
+     * The default timezone
+     */
+    @WithDefault("UTC")
+    String timeZone();
```
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java]: I would make it optional and use System as default, any reason for choosing UTC by default?

**Review by ia3andy — APPROVED** [review]
Awesome!! This is great stuff :)

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java (hunk 4 above)
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java]:
```suggestion
    @ConfigDocDefault("document timezone if provided or system timezone")
    Optional<String> timeZone();
```

File: roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java [hunk 5]
```diff
@@ -49,9 +50,16 @@ public interface RoqFrontMatterConfig {
...
      * Format for dates
      */
-    @WithDefault("yyyy-MM-dd HH:mm:ss Z")
+    @WithDefault("yyyy-MM-dd[ HH:mm][:ss][ Z]")
     String dateFormat();
 
+    /**
+     * The default timezone
+     */
+    @WithDefault("UTC")
```
Comment (ia3andy) [inline roq-frontmatter/deployment/src/main/java/io/quarkiverse/roq/frontmatter/deployment/RoqFrontMatterConfig.java]: When using optional, no need to add @WithDefault
```suggestion
```

**Review by melloware — APPROVED** [review]

**Pull request merged**